### PR TITLE
ext_authz: make the ext_authz filter a dual filter

### PR DESCRIPTION
**Pull request by pianiststickman**
This is a revival of #25535 with changes for previous review comments.

Risk level: low
Testing: integration tested
Docs changes: n/a
Release notes: makes the ext_authz filter a dual filter.

See also: #23071 (model), #10455

### CODE REVIEW

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (alyssawilk) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:82]: Let's config enforce this and add a test. 

I think we can just add a boolean for upstream vs downstream filter in the constructors of DualInfo and then add this to the checks in config.cc

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (alyssawilk) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:264]: After we config-validate we can add a comment here that clearRouteCache should onyl be true if  decoder_callbacks_->downstreamCallbacks() is also true

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (alyssawilk) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:111]: Looks to me like you almost nailed the requirements for source/docs/upstream_filters.md
the one I think is outstanding is 

Any code accessing the downstream connection checks to make sure it is present.

I see a bunch of decoder_callbacks_->connection() which I _think_ won't work for shadow requests.  
I think we should put in connection checks, add a test of shadowing that would have crashed without those changes, and comment the limitation in the authz filter docs.

**Review by alyssawilk — COMMENTED** [review]
Very nice change - I know a lot of folks are excited for this particular one to be available as an upstream filter so thanks for working to make it happen!
/wait

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (pianiststickman) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Done.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (pianiststickman) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Done.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (pianiststickman) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:265]: Done.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (pianiststickman) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:82]: Done.

**Comment by alyssawilk** [comment]
/retest
/wait on CI
(please remove waiting label if retest fixes things, but looks like several are failing so I assum eit's real?)

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (pianiststickman) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Fixed - it turns out ext_authz filter calls into a common `createHttpCheck()` that also references a bunch of things on the downstream connection, like the peer certificate and the TLS session, so I've also config-disallowed those fields from being set on the upstream filter chain.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:140]: I think this is something that may be of interest to ext_authz server. Is there a specific reason why this is not going to work for upstream ext_authz?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (yanavlasov) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:271]: Yes, we can do it. We can add `security_posture_for_upstream_filter` to make it different from the downstream filter posture.

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (pianiststickman) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:140]: It, and other things, seem to require the downstream connection, which is not available on upstream filters. source/extensions/filters/common/ext_authz/check_request_utils.cc

**Comment by pianiststickman** [comment]
/retest

AFAICT current failures are not related.

**Review by alyssawilk — COMMENTED** [review]
This largely LGTM but I do want to sort out the API question before landing as it may limit our options given API rules.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (ggreenway) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:140]: It seems like there should be some way to get it. Maybe look at how xfcc works, because that includes the cert in upstream requests.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (ggreenway) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:81]:
```suggestion
  // It is an error to set this field when the filter is configured on an upstream filter chain.
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
I think this needs some docs. What is the behavior when this is configured as an upstream filter? When would one configure this instead of a downstream auth filter?

**Comment by pianiststickman** [comment]
> I think this needs some docs. What is the behavior when this is configured as an upstream filter? When would one configure this instead of a downstream auth filter?

We have a use case where ext_authz depends on some dynamic metadata to have been filled in by a previous filter in the filter chain (it's shipping that metadata off to the Check server), but that previous filter happens to depend on cluster/endpoint selection so it is an upstream filter. More broadly, one could imagine that the authz filter could ship cluster metadata to the Check server and would thus depend on cluster selection having happened already. The filter otherwise behaves identically to the downstream case, this is enabling the ability to put it on an upstream filter chain.

> It seems like there should be some way to get it. Maybe look at how xfcc works, because that includes the cert in upstream requests.

At the moment, filters in an upstream filter chain do not have access to the downstream connection. XFCC is being implemented in the HttpConnectionManager which does have references to the downstream connection. I'm proposing disallowing this combination until the plumbing to enable this is present.

I'm happy to modify any additional docs as requested.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (pianiststickman) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto]: XFCC is being implemented in the HttpConnectionManager which has references to the downstream connection. The upstream filter chain, at the moment, does not.

I am reading alyssawilk@'s comment in https://github.com/envoyproxy/envoy/pull/29173#discussion_r1307782625 as saying that we could potentially plumb this into upstream requests. It being that this isn't currently being done, I'm proposing disallowing this combination until such time as downstream connection fields are available on the upstream streamInfo.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (alyssawilk) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:141]: OK confirmed with Adi offline we're fine with 1->3 so let's go with that (resolving this)

File: source/extensions/extensions_metadata.yaml
Comment (alyssawilk) [reply source/extensions/extensions_metadata.yaml:311]: Final word here - I'd like to have https://github.com/envoyproxy/envoy/issues/29305 sorted before we land this so we can tag the upstream use as alpha.

Alternately if you can get API shephard sign-of adding a ::warning to the proto that upstream use is alpha (see https://github.com/envoyproxy/envoy/issues/29305) for tooling I think that's an OK thing in the interim.  Bonus points if you can comment inline here as well but I don't recall if the script allows.

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo the two remaining comments.  @lizan can you do an owner pass?

**Comment by alyssawilk** [comment]
/wait on main merge and API warnings

**Comment by alyssawilk** [comment]
Ok, with https://github.com/envoyproxy/envoy/pull/29488/files merged if you merge main  you should be ablle to have upstream authz tagged as alpha.  Also needs main merge

**Review by alyssawilk — DISMISSED** [review]

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (pianiststickman) [reply test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc]: Done.

**Comment by pianiststickman** [comment]
@ggreenway PTAL, thanks!

(AFAICT the presubmit failures are not related but they seem awfully flaky...)

**Review by alyssawilk — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
